### ISSUE TITLE
Approach for refreshing s3contents/s3fs (AWS) keys?

### ISSUE THREAD

**Issue by milutz** [danielfrg/s3contents#79]
Using new features in Jupyter 1.0.0 I've been able to build a JupyterHub config that will automatically update the keys in AWS credentials file for each user (so I can switch to safer expiring temporary keys, instead of scary permanent keys).

It appears, unfortunately, that s3contents (likely because of unhelpful behaviors of s3fs/boto) doesn't ever reread the auth info for a give Jupyter instance.

@danielfrg  If I wanted to make s3contents reload the keys (which guessing may require instantiating a new s3fs object) either based off a timeout or based off of getting an access error would you have a recommendation on how I would approach that?

It looks like https://github.com/danielfrg/s3contents/blob/master/s3contents/s3_fs.py is where the s3fs library is consumed, maybe adding a method that repeats the `__init__` `self.fs = s3fs.S3FileSys...` call ... then ... hmm..

Any ideas?

**Comment by danielfrg** [comment]
Yeah, I don't see any issue with that approach, a method `create_fs` that moves some code from `__init__` that creates the `s3fs` object and its called from the `__init__`. Are you interested on making a PR for that?

**Comment by milutz** [comment]
@danielfrg Yes - I'm just looking for guidance on the right way to put it together.   Assuming I get something working I'll PR it back to the project. 

What would be your thought on the best way to trigger the refresh?  Is there a systemwide error handler I could highjack to call the refresh code? (there aren't any timer systems already built in are there?)

(I'm still working on #78 too, hoping to get that to a push sometime soon too)

**Comment by milutz** [comment]
After attempting the above and having it not refresh the keys, I did a deep dive into boto3/botocore and found that there _is_ a way to have boto itself do the key refresh.

I tested this solution by doing a few overrides in s3contents and S3FS ("S3FS" is the s3contents code - not the dash library), and it seems to work cleanly.

So instead of making PR for what I wrote when I opened this issue, I'm going to work on building a PR for letting you predefine the boto3 object when you do the calls to startup s3contents.

**Comment by peter-friedland-bose** [comment]
Did y'all get something to work?  I'm currently using AWS environment variables with session token and also need to refresh the tokens periodically. Seemingly neither Jupyter or S3Contents pays any mind to my changes to environment variables after the Jupyter is running. Are you thinking of doing something like is described in this ancient botocore issue? 
https://github.com/boto/botocore/issues/704

**Comment by milutz** [comment]
@peter-friedland-bose Yes I have something working, and yes I did it more-or-less as talked about in boto/botocore#704

I'm working my way up to doing a PR with my changes which are all about allowing you to push a boto3 session into s3contents - through which you can pre-create a refreshing boto3 session as listed in above.

You can look at https://github.com/milutz/s3contents/tree/patch-for-passing-botocore if you want to see what I'm cleaning up to make a PR with.  (I'm uncertain if I'm building some of the class structures correctly, so I won't be surprised if there will be a bit of back-and-forth before @danielfrg accepts the PR)

I'll try and get something written up in a day for a first-pass submission (I'll package up the bit of the jupyter/jupyterhub config needed too)

**Comment by peter-friedland-bose** [comment]
@milutz I'd be happy to test it out or help in any way.  I'm looking at your changes now.  Please let me know.

**Comment by milutz** [comment]
@peter-friedland-bose Here is a super high level description of what I'm doing:

I'm running a single host JupyterHub, whose config is setup to create docker based Jupyters, which use s3contents for all their s3 access 

(this change could work for multi-host solutions too but something beyond docker drive mounts would be needed to forward the keys - or the re-gen logic would have to be contained on the jupyter client, which isn't the behavior I want)

1. JuptyerHub central app is generating refreshed per-user AWS keys and writing them into per-user directories (JupyterHub 1.0 provides some machinery to do this, but takes some extra glue code too)
2. When JupyterHub spins up a user it creates a docker container that mounts _just_ that users key directory in the magic path that boto3 expects key files to live
3. The Jupyter config then makes a refreshing boto3 instance (one that will reread the config) and uses that to init my changed s3contents
4. my mod to s3contents adds a `c.S3ContentsManager.init_s3_hook` that can have the function that creates the refreshing boto3 instance and forwards that through s3contents into dash/s3fs (which is what really does the S3 heavy lifting)

**Comment by peter-friedland-bose** [comment]
@milutz I'm doing something similar but simpler, I think.  Instead of JupyterHub, I'm spinning up  standalone Jupyter Lab in a Docker container(s) on a per user basis.  When the container is started, I pass in environment variables including the standard AWS ones containing STS generated temporary tokens. My container's s3contents config code grabs the environment variable contents when Jupyter Lab starts. Subsequently, I will refresh the user(s) temporary tokens and store them externally and the container will query them periodically to refresh and avoid expiration.  I'm guessing I will have to use a standard credentials file instead of environment variables with this solution and update the contents ~/.aws/credentials with new tokens periodically. 

Am I getting this right?

**Comment by milutz** [comment]
@peter-friedland-bose Yup that all sounds completely reasonable to me.

The only difference is that you have your own system (your case) in place of jupyterhub (my case)

In the case of docker (unless you've figure out something I haven't...) environment vars are completely static once the container is started up so if you want to update information for a running container, systems like docker mounts are a good way to go

I do believe there are types of STS tokens that have the right to refresh themselves.  If your ok with that style token then you would also have the option of having the refresh code do the refresh itself, and not require anything new be passed in

**Comment by milutz** [comment]
@peter-friedland-bose take a look at this https://dev.to/li_chastina/auto-refresh-aws-tokens-using-iam-role-and-boto3-2cjf (its one of the sources that I used to figure out how to get my solution to work.

If you break down that solutions `_refresh` I believe you'll find that its using its active temp keys to call `self.sts_client.assume_role(**params).get("Credentials")` before they expire - I.E. just endlessly refreshing an instance of assume_role

If your good with that, its likely a simnifically easier path (and you can just copy the code used in that blog to do it)

My changes to s3contents are still needed, but then you wouldn't have to mess with docker mounts or external engines that key refreshing the keys

**Comment by peter-friedland-bose** [comment]
@milutz I'm having difficulty implementing the refresh the temporary tokens concept and I've basically followed the code pattern exactly as the article explains.  I start with a set of temporary credentials that are set to the standard AWS environment variables.  I use the _refresh method pattern as the article shows.  After 15 min (expiry for my tokens), I'm getting expired token messages. While I didn't dig into the botocore code to see exactly how this works, I sense that the _refresh method is simply using the tokens in the initial ENV variables which I'm assuming is why I don't get the refresh.

**Comment by milutz** [comment]
@peter-friedland-bose I would recommend using the style of debug in my example above.

In your jupyter config include `from tornado.log import access_log` and then add debug messages like `access_log.debug("key refresh called, pulled key: " + str( config['default']['aws_access_key_id']))`.

If you add enough messages then you can see which key is being refreshed, and what the outcome is - certainly took me a few tries before I got it right

**Comment by milutz** [comment]
@peter-friedland-bose if you post your jupyter config (at least the s3contents relevant part of it) and the logs I can try and help you dig through it too

**Comment by peter-friedland-bose** [comment]
@milutz I was simply doing the refresh outside of Jupyter to see how it works.  Question... should I be using an AWS credentials file rather than ENV variables when trying to refresh temp credentials?  I'll send my code too.

**Comment by milutz** [comment]
@peter-friedland-bose the way you push in the credentials in shouldn't matter assuming your not getting an error when `sts_client.assume_role(**params).get("Credentials")` is originally called.

(My number one question would be _if_ `sts_client.assume_role` in the `_refresh` method is being called - if the parts aren't wired together right it will never get called to do the refresh)

**Comment by peter-friedland-bose** [comment]
@milutz I lied.  The error for the code above is not in the _refresh method. I could swear in the many tests I ran that it was dying in _refresh
```
Traceback (most recent call last):
  File "refresh-tokens.py", line 55, in <module>
    print(json.dumps(sts_client.get_caller_identity(), default=str, indent=4))
  File "/Users/pfriedland/Library/Python/3.7/lib/python/site-packages/botocore/client.py", line 357, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/Users/pfriedland/Library/Python/3.7/lib/python/site-packages/botocore/client.py", line 661, in _make_api_call
    raise error_class(parsed_response, operation_name)
botocore.exceptions.ClientError: An error occurred (ExpiredToken) when calling the GetCallerIdentity operation: The security token included in the request is expired
```

**Comment by milutz** [comment]
@peter-friedland-bose Its taken a while to setup a test env for you use case (because I'm not using roles) but I suspect the issue is that your not using the sts_client instance connected to the RefreshableCredentials object (your code has `response = sts_client.assume_role(**params).get("Credentials")` in _refresh, the blog example has `response = self.sts_client.assume_role(**params).get("Credentials")` (note the `self.sts_client`)

I'm building my way up to a test to prove it.

**Comment by peter-friedland-bose** [comment]
I really, really appreciate your help.  I was confused by the reference to 'self' in _refresh method as you mention. I will also test your suggestion ...

**Comment by milutz** [comment]
@peter-friedland-bose So in your code I changed the _refresh block to:

```python
class mytest:
    def _refresh(self):
        " Refresh tokens by calling assume_role again "
        self.sts_client= boto3.client('sts', aws_access_key_id=self.credentials["access_key"],
                                             aws_secret_access_key=self.credentials["secret_key"],
                                             aws_session_token = self.credentials["token"]) 
        response = self.sts_client.assume_role(**params).get("Credentials")
        print(str(response))
        credentials = {
            "access_key": response.get("AccessKeyId"),
            "secret_key": response.get("SecretAccessKey"),
            "token": response.get("SessionToken"),
            "expiry_time": response.get("Expiration").isoformat(),
        }
        self.credentials = credentials
        return credentials

miketest = mytest()
miketest.credentials = metadata
miketest.params = params
```

And I changed the session_credentials block to:

```python
session_credentials = RefreshableCredentials.create_from_metadata(
    metadata=metadata,
    refresh_using=miketest._refresh,
    method="sts-assume-role",
)
```
I'm _sure_ I'm not implementing this as cleanly as the original blog author did (at minimum the `mytest` and `miketest` class and vars could have better names) - but I'm also a bit confused how the code the blog author posted could have worked (their examples definitely imply things are inside of a class, but I don't see which class)

Let me know if you want anything explained in more detail - its all about moving the `sts` response into an object that has the _refresh code so when _refesh is next run, that it has the current cert. not the original cert (which expires at 15 mins as noted)

I haven't run this quite long enough to know for sure that it is working, will update in a bit

**Comment by peter-friedland-bose** [comment]
@milutz  I see what you are doing. I was studying the botocore code for RefreshableCredentials  class.  I agree  my original _refresh implementation always uses my starting credential tokens and not the ones that are refreshing within the RefreshableCredentials object at each refresh.  I'm going to run it too.  Very cool and thank you again.

**Comment by peter-friedland-bose** [comment]
@milutz Yep.  It has kept going past the original credentials 15 min expiration.  Good deal.
I understand what is going on with RefreshableCredentials now.  
Next, your enhanced s3Contents will somehow use the refreshed credentials as long as keep calling in a run loop in my jupyter container.  Or at least that is what I had in mind.  I'll wait to see what you have in mind. Thanks again and have a good evening buddy.

**Comment by milutz** [comment]
@peter-friedland-bose Awesome!

Ok, last step:
 
Assuming your working in docker you need to remove whatever line of your `Dockerfile` that installs s3contents and (until my changes are merged) put something like:
```
RUN git  clone   https://github.com/milutz/s3contents.git --branch patch-for-passing-botocore
WORKDIR s3contents
RUN python setup.py install
WORKDIR ..
RUN rm -rf s3contents
```

Then, in your `jupyter_notebook_config.py` you are going to bundle up your previous work, with the part that makes the refreshing session bundled in a function, something like:
```
def make_key_refresh_botocore(self):
    refresh_session =  get_session() # from botocore.session
    refresh_session._credentials = session_credentials
    my_s3_session =  boto3.Session(botocore_session=refresh_session)
    self.boto3_session = my_s3_session
```

And then add a config line to connect that function into my new code:
```
c.S3ContentsManager.init_s3_hook = make_key_refresh_botocore
```
At this point you should have you working!
(Mind you this only deals with the file browser part of juptyer, you'll have to do some more work if you want to access these same keys _inside_ of a notebook)

**Comment by peter-friedland-bose** [comment]
@milutz Excellent.  I will try out later today. Thank you again.  I learned a lot here.

**Comment by peter-friedland-bose** [comment]
@milutz Looks like I don't need to pull the branch - looks all merged into master.  I got distracted building more stuff into my now huge container so I'm just getting around to the neat refresh stuff.  Thanks again Mike !!  Maybe you can coach me on how to get the credential variables generated in my _refresh method into notebook variables.  I need to use them in Spark Session property values.

**Comment by milutz** [comment]
@peter-friedland-bose The quick answer is its just the same code, but baked into a notebook - though in your case your going to have pick a method of how the notebook will have access to an active copy of the keys to get started (as they are just memory resident inside of s3contents (... its easy for me because the method I'm using requires that the key live in the filesystem already)

**Comment by peter-friedland-bose** [comment]
@milutz So when the container starts, AWS_XXX environment variables with temp tokens are passed in.  But env variable tokens expire some hours later.  Is there a way, you know of, to share the dynamically updated tokens generated from the _refresh method into "magic" variables in a notebook or something?  I'm probably talking outta my ear, eh?

**Comment by milutz** [comment]
Ooo, that would be clever wouldn't it --- I don't really know anything of that code, but my guess would be "that's hard" because (I think) the magics are tied to the interpreter-kernal, not the jupyter-UI/UI-backend, which is where s3contents lives.

@danielfrg Do you know if there is any easy way to expose vars living inside of s3contents to the notebook (or notebook UI)?

**Comment by peter-friedland-bose** [comment]
@milutz - maybe I should simply write the refreshed tokens into ~/.aws/credentials file format and read that as needed.  Then boto3 should see them too.  As long as my environment variables aren't named the same as the "well known" AWS ones.

**Comment by danielfrg** [comment]
That's right, those are a different process running and they don't share any env variables. There might be ways to pass them but I don't think we would want to do that for the credentials. 
I would recommend to just read them from a central place as you mentioned in the last comment.

**Comment by milutz** [comment]
Given a fix to this has been merged (pull request #81 ) and released, seems like this should be closed too :)
